docs/source/meta_init.rst

```diff
@@ -0,0 +1,239 @@
+    >>> # Do some form of inspection (e.g. apply a sharding algorithm).
+    >>>
+    >>> torch.nn.utils.materialize(m)
+    >>> m.weight
+    Parameter containing:
+    tensor([[-1.4677e+24, 4.5915e-41, 1.4013e-45, 0.0000e+00,
+             -1.4677e+24, 4.5915e-41]], requires_grad=True)
+
+
+You can also materialize just one or more submodules of a large model.
+
+::
+
+    >>> import torch
+    >>>
+    >>> with torch.nn.utils.meta_init():
+    ...     m = MyVeryLargeModel()
+    >>>
+    >>> # Do some form of inspection (e.g. apply a sharding algorithm).
+    >>>
+    >>> # Only materialize `sublayer1` and `sublayer2`.
+    >>> torch.nn.utils.materialize(m.sublayer1)
+    >>> torch.nn.utils.materialize(m.sublayer2)
+
+
+Note that :func:`meta_init` overrides even explicitly passed ``device``
+arguments and forces the use of the meta device.
+
+::
+
+    >>> class MyModule(torch.nn.Module):
+    ...     def __init__(self):
+    ...         super().__init__()
+    ...         self.param = torch.nn.Parameter(torch.ones([3], device="cpu"))
+    ...
+    >>> with torch.nn.utils.meta_init():
+    ...     m = MyModule()
+    >>> m.param
+    Parameter containing:
+    tensor(..., device='meta', size=(10, 10), requires_grad=True)
+    >>>
+    >>> torch.nn.utils.materialize(m)
+    >>> m.param
+    Parameter containing:
+    tensor([1., 1., 1.], requires_grad=True)
+
+
+Although the main use case of :func:`meta_init` is large module materialization,
+its underlying machinery is actually more flexible. You can also construct and
+later materialize individual tensors.
+
+::
+
+    >>> with torch.nn.utils.meta_init():
+    ...     t1 = torch.ones([3])
+    ...     t2 = t1 + 2
+    >>> t1
+    tensor(..., device='meta', size=(3,))
+    >>> t2
+    tensor(..., device='meta', size=(3,))
+    >>> # By default, `materialize()` clears the meta-init cache. If you want
+    ... # to materialize more than one object, set the `keep_cache` parameter
+    ... # to `True`.
+    >>> torch.nn.utils.materialize(t1, keep_cache=True)
+    >>> t1
+    tensor([1., 1., 1.])
+    >>> torch.nn.utils.materialize(t2)
+    >>> t2
+    tensor([3., 3., 3.])
+
+
+API
+---
+.. autosummary::
+    :toctree: generated
+    :nosignatures:
+
+    meta_init
+    is_meta_init
+    materialize
+    clear_meta_init_cache
+
+.. autofunction:: meta_init
+.. autofunction:: is_meta_init
+.. autofunction:: materialize
+.. autofunction:: clear_meta_init_cache
+
+.. _common-failures:
+
+Common Failure Cases
+--------------------
+Since meta initialization relies on constructing modules and their tensors on
+the meta device, there are some certain usage patterns that will fail when used
+with :func:`meta_init`.
+
+**A module that uses an operator not supported by the meta backend:**
+If the module uses an operator that is not yet supported by the meta backend,
+the operator call will fail. If you experience such failure, please open a
+GitHub issue and let us know which operator you are using.
```

Comment:
If this ends up becoming a huge list of operators, I was wondering if we can provide an easy pythonic way for users to write their own meta operators to fix issues like these.

Reply:
I think we are getting there. There is an ongoing process to port our regular operators to structured kernels which support the meta backend out of the box. For instance I ported `tril` and `triu` in #67055 to unblock some HG models, along with #67032 that enables meta support in several tensor factories.

Having said that if the user wants to have their own "fix", I think the most pythonic and cleanest way would be to implement a `__torch_dispatch__` function. Let me experiment with that approach, if I see that it is feasible, I can include it in the doc.

Reply:
cc @bdhirsh for meta tensor coverage. @bdhirsh I remember you said there was was some unsanctioned way to add meta tensor coverage without refactoring code into a structured kernel? Could that be a workaround?

Reply:
There isn't really, except for explicitly writing and registering the meta kernel (like with `arange`, [here](https://github.com/pytorch/pytorch/blob/07932e27356c32df8a3c17361e4779c635d2d8ce/aten/src/ATen/native/native_functions.yaml#L596)). So far we've been only doing that for ops that we can't port to structured though, and not as a temporary workaround.